[ENG-1132] Add Evaluation documentation pointing to Literal cookbooks (#25)
Chainlit/literal-docs: concepts/evaluation.mdx

title: "Evaluation"
---

## Introduction

Evaluation is key to enable continuous deployment of LLM-based applications and guarantee that newer
versions perform better than previous ones. To best capture the **user experience** one must understand the
multiple steps which make up the application. As AI applications grow in complexity, they tend to chain 
multiple steps.

> Literal decorators let you catch the various steps of an AI application and gather the inputs
> and outputs of each. With those, one may evaluate the performance of their application at any level, 
> building the most relevant KPIs for end-users:
> - Thread &mdash; user satisfaction
> - Agent Run &mdash; task completion
> - LLM Generation &mdash; hallucination

Reviewer: ```suggestion
Literal lets you log & monitor the various steps of your LLM application. By doing so, you can continuously improve the performance of your LLM system, building the most relevant metrics:
- Thread &mdash; user satisfaction, etc.
- Agent Run &mdash; task completion, etc.
- LLM Generation &mdash; hallucination, etc.
```

An example is the vanilla Retrieval Augmented Generation (RAG), which augments Large Language Models
(LLMs) with domain-specific data. Examples of metrics you can score against are:

> - context relevancy
> - faithfulness
> - toxicity

Reviewer: ```suggestion
- context relevancy
- faithfulness
- answer relevancy
```

- etc.

<Frame caption="Vanilla RAG Thread">
  <img src="/images/rag-thread.png" alt="A thread on a RAG application" />
</Frame>

> Literal offers a compelling user/developer interface to build and interact with datasets.
> Through `Dataset` objects, users can iterate on an application parameter and view
> its impact via Experiments. 

Reviewer: ```suggestion
Literal offers a compelling user/developer interface to build and interact with datasets.
Through `Dataset` objects, users can iterate on a parameter of the LLM system - such as prompt template, LLM provider, temperature, etc - and view its impact via Experiments. 
```

> For instance, a developer can iterate on their prompt template and check the impact of their
> modifications on production data. 

Reviewer: ```suggestion
A developer can also iterate on their prompt template and check the impact of their
modifications on production data. 
```

Evaluations can either take place:
- on a `Dataset` which contains a number of steps' inputs/outputs to score &mdash; these are **offline evaluations**
- on live production data &mdash; those are **continuous evaluations**

## Offline evaluations

With offline evaluations, developers can leverage their favorite framework, such as
promptfoo, Ragas, OpenAI Evals, LangChain Evaluators, etc.

> Our [Prompt Iteration with Promptfoo](https://github.com/Chainlit/literal-cookbook/tree/main/evaluation/typescript/prompt-iteration-promptfoo)
> cookbook shows how a developer can leverage the `promptfoo` library in TypeScript to
> compute similarity scores and persist the results of their experiments directly on Literal. 
> 
> For Python aficionados, our [Context Relevancy with Ragas](https://github.com/Chainlit/literal-cookbook/tree/main/evaluation/python) notebook experiments on an example RAG
> application to score context relevancy.

Reviewer: ```suggestion
- [Prompt Iteration with **Promptfoo**](https://github.com/Chainlit/literal-cookbook/tree/main/evaluation/typescript/prompt-iteration-promptfoo):
Our cookbook shows how a developer can leverage the `promptfoo` library in TypeScript to
compute similarity scores and persist the results of their experiments directly on Literal. 
- [Context Relevancy with **Ragas**](https://github.com/Chainlit/literal-cookbook/tree/main/evaluation/python): our notebook experiments on an example RAG
application to score context relevancy.
```
